### PR TITLE
[EM-390] Enhance Magento plugin to support FPX

### PR DESCRIPTION
**Pull request by Ashraf-alkibsi**
#### 1. Objective

As a merchant developer, I want an ability for Magento plugin to make payments via FPX so that I can have payment options provided to my customers.

**Related information**:
Related issue(s): #< EM-390> 

#### 2. Description of change

Added the code for FPX payment method. and refactor the TruemoneyDataAssignObserver.php & ConveniencestoreDataAssignObserver.php to use OffsiteDataAssignObserver to remove duplicated codes

#### 3. Quality assurance

Specify where and how you tested this and what further testing it might need.
tested by purchasing some products locally. still need to be tested on staging.  testing installment payment and Truemoney might be needed as well. 

**🔧 Environments:**

Specify the details of your test environments, including, for each, the platform version (on which the plugin was run), the Omise plugin version, and the versions of your system software such as PHP or Ruby.

i.e.
- **Platform version**: Magento CE 2.1.5.
- **Omise plugin version**: Omise-Magento 2.1.
- **PHP version**: 7.0.16.
was tested on magento docker https://git.omise.co/omise/Magento-docker
- **Platform version**: Magento CE 2.2.
- **Omise plugin version**: Omise-php 2.13
- **PHP version**: 7.1.

**✏️ Details:**

Explain how to manually test this feature.
For example if changes were made in the UI or in the API, explain where and if any specific access is needed.

#### 4. Impact of the change

List the steps that must be taken for this PR to work.
E.g.: rake yak:shave, Add "yak_key" to environment variables, ...
the country and the currency of the system should be set to Malaysia and Malaysian ringgit

Be sure to include all systems that needs to be changed or which system is affected by the change
(Ex: Requires Elastic search to be installed and configured in secrets.yml).

Note: Please provide a screenshot if your changed impact to UI.
in customers view
Screen Shot 2021-06-04 at 11 23 23 AM

and in admin view

Screen Shot 2021-06-04 at 11 46 40 AM

#### 5. Priority of change

Normal, High or Immediate.
Normal

#### 6. Additional Notes

Any further information that you would like to add.

### CODE REVIEW

**Comment by kan98** [comment]
Acceptance test: passed ✅ 
We discussed for you to handle errors when capability doesn't fetch the banks list and it works great, so thanks for adding that.

One feedback from the test that I have is the appearance of the banks list. When we click on a bank, the highlighted box leaves some white space on the left. Could you add a negative margin on the `<option>` or remove the left padding on `<select>` and add that padding on `<option>` instead? Also can you add `cursor: pointer` to the banks list so the mouse becomes a pointer when the user selects the banks lists. Looks great otherwise!

**Comment by kan98** [comment]
> Acceptance test: passed ✅
> We discussed for you to handle errors when capability doesn't fetch the banks list and it works great, so thanks for adding that.
> 
> One feedback from the test that I have is the appearance of the banks list. When we click on a bank, the highlighted box leaves some white space on the left. Could you add a negative margin on the `<option>` or remove the left padding on `<select>` and add that padding on `<option>` instead? Also can you add `cursor: pointer` to the banks list so the mouse becomes a pointer when the user selects the banks lists. Looks great otherwise!

All this is fixed now, just tested. ✅

File: view/frontend/web/template/payment/offsite-fpx-form.html [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+                        <option data-bind="attr:{value: code,
+                         class:code}, html:'<span>'+name+'</span>'">
+                        </option>
+                    </select>
+                </div>
+            </fieldset>
+        </form>
+        <div class="actions-toolbar">
+            <div class="primary">
+                <button class="action primary checkout" type="submit"
```
Comment (kan98) [inline view/frontend/web/template/payment/offsite-fpx-form.html:72]: can we disable this until a bank is selected? The payment will fail anyways but nice to have some client side validation

**Review by kan98 — APPROVED** [review]

File: Gateway/Request/APMBuilder.php [hunk 2]
```diff
@@ -16,6 +17,7 @@
 use Omise\Payment\Model\Config\Truemoney;
 
 use Omise\Payment\Observer\ConveniencestoreDataAssignObserver;
+use Omise\Payment\Observer\FPXDataAssignObserver;
```
Comment (danfowler) [inline Gateway/Request/APMBuilder.php:20]: Very small, but the capitalization approach for FPX here (`FPX`) is different than the one above (`Fpx`). `FpxDataAssignObserver` would be more consistent.

File: etc/adminhtml/system.xml [hunk 3]
```diff
@@ -334,6 +334,35 @@
...
+                    <field id="title" translate="label" type="text" sortOrder="420" showInDefault="1" showInWebsite="1" showInStore="1">
+                        <label>Title</label>
+                        <comment>This controls the title which the user sees during checkout.</comment>
+                        <config_path>payment/omise_offsite_fpx/title</config_path>
+                    </field>
+                    <field id="allowspecific" translate="label comment" type="select" sortOrder="421" showInDefault="1" showInWebsite="1" showInStore="1">
+                        <label>Allowed Countries</label>
+                        <config_path>payment/omise_offsite_fpx/allowspecific</config_path>
+                        <source_model>Magento\Payment\Model\Config\Source\Allspecificcountries</source_model>
+                        <comment>If set to specific, guest customers will not have a billing country and may not be able to checkout.</comment>
```
Comment (danfowler) [inline etc/adminhtml/system.xml:355]: Can this be rephrased? e.g. "If specified, ..."

Comment (Ashraf-alkibsi) [reply etc/adminhtml/system.xml:355]: Should i change this for other payments method as well?

Comment (danfowler) [reply etc/adminhtml/system.xml:355]: I think so, but I'm going to amend my suggestion to "If set to specific countries, ..."

Comment (danfowler) [reply etc/adminhtml/system.xml:355]: Let's confirm the wording tomorrow morning first

File: view/frontend/web/template/payment/offsite-fpx-form.html (hunk 1 above)
Comment (Ashraf-alkibsi) [reply view/frontend/web/template/payment/offsite-fpx-form.html:72]: yes, it is good to have it. I added it now 
`enable: (getCode() == isChecked() && selectedFpxBank)`

File: etc/adminhtml/system.xml (hunk 3 above)
Comment (danfowler) [reply etc/adminhtml/system.xml:355]: I think the best wording in the end is "If not set to all, ..."

**Review by danfowler — APPROVED** [review]

**Pull request merged**